datafusion/functions/src/string/concat_ws.rs

```diff
@@ -46,9 +45,8 @@ impl Default for ConcatWsFunc {
 
 impl ConcatWsFunc {
     pub fn new() -> Self {
-        use DataType::*;
         Self {
-            signature: Signature::variadic(vec![Utf8], Volatility::Immutable),
+            signature: Signature::variadic_any(Volatility::Immutable),
```

Comment:
@alamb having a few issues with the `signature` here. I'm assuming that `variadic` will allow for a vector of common types but it effectively will coerce the other types in to whatever the first type is? In this case the separator is generally always a `Utf8` type which then causes the subsequent parameters to be cast to a utf8.

```
    pub fn new() -> Self {
        use DataType::*;
        Self {
            signature: Signature::variadic(
                vec![Utf8, Utf8View, LargeUtf8],
                Volatility::Immutable,
            ),
        }
    }
```

Is there something similar to: 

`Exact(Utf8, variadic([utf8, utf8view, largeutf8]))` ?

Looking through the type signatures I don't see an established way to do that (that I'm aware of). Any suggestions?

Reply:
> Looking through the type signatures I don't see an established way to do that (that I'm aware of). Any suggestions?

One way to do this is to explicitly list all the valid combinations using [`TypeSignature::OneOf`](https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.TypeSignature.html#variant.OneOf)

for example, 
https://github.com/apache/datafusion/blob/e8ac93a0ffd490bd549872366bf7fd1b77ffd7f8/datafusion/functions/src/string/btrim.rs#L56-L62